File: 0019_remove_nth_node_from_end_of_list/solution.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        count: int = 0
        prev: ListNode = None
        curr: ListNode = head

        while curr:
            curr = curr.next
            count += 1

        if count == n:
            return head.next

        curr = head
        i: int = 0
        while i < count - n:
            prev = curr
            curr = curr.next
            i += 1

        prev.next = curr.next

        return head

        

File: 0019_remove_nth_node_from_end_of_list/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// runtime: 100.00%, memory: 7.94%
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode list = new ListNode(0, head);
        ListNode temp = list;

        for (int i = 0; i < n; i++) {
            head = head.next;
        }

        while (head != null) {
            head = head.next;
            temp = temp.next;
        }

        temp.next = temp.next.next;

        return list.next;
    }
}